sdks/java/harness/src/main/java/org/apache/beam/fn/harness/debug/DataSamplingDescriptorModifier.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.debug;
+
+import org.apache.beam.fn.harness.ProcessBundleDescriptorModifier;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+
+/**
+ * Modifies the given ProcessBundleDescriptor by adding a DataSampling operation as a consumer to
+ * every PCollection.
+ */
+public class DataSamplingDescriptorModifier implements ProcessBundleDescriptorModifier {
```

Comment:
I was thinking that you would update PCollectionConsumerRegistry#MetricTrackingFnDataReceiver` and `PCollectionConsumerRegistry#MultiplexingMetricTrackingFnDataReceiver` and add an `if(sampling) { doSampling }` to them. The `if` should be really cheap since the value should always be false or true so the JVMs ability to predict the branch should be very high. Alternatively you could also create a sub-class of the three receivers there to add sampling as necessary.

You can validate the `if` approach by running the JMH benchmark:
```
./gradlew -info :sdks:java:harness:jmh:jmh -Pbenchmark=ProcessBundleBenchmark.testLargeBundle
```

Reply:
Gotcha, that makes sense. I changed to the suggested implementation and ran the benchmark:

**Benchmark with this PR (data sampling not enabled)**
Benchmark                                Mode  Cnt     Score     Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1640.020 ± 236.240  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 3,5,main]) completed. Took 5 mins 9.573 secs.

**Benchmark with this PR (data sampling enabled)**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1094.943 ± 99.748  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 4,5,main]) completed. Took 5 mins 9.603 secs.

**Benchmark at upstream/master**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1808.108 ± 30.062  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 5,5,main]) completed. Took 5 mins 9.064 secs.


Clearly, there's some environmental factors affecting the benchmarking. But, I think it can at least be said this PR with data sampling disabled won't completely hinder performance.

Reply:
Redone with atomics:

**yes data sampling (no atomics)**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1150.680 ± 25.650  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 7,5,main]) completed. Took 5 mins 9.453 secs.

**yes data sampling (with atomics)**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1547.291 ± 50.887  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 6,5,main]) completed. Took 5 mins 8.804 secs.


**no data sampling**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1697.123 ± 74.466  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 3,5,main]) completed. Took 5 mins 8.911 secs.

Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1709.427 ± 38.715  ops/s
:sdks:java:harness:jmh:jmh (Thread[included builds,5,main]) completed. Took 5 mins 8.804 secs.

**At head**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1807.738 ± 15.295  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 2,5,main]) completed. Took 5 mins 8.921 secs.

Reply:
This looks much better since the variation in the runs is smaller but 5% is still not small.

Hopefully marking the fields final will address that but if not we might want to go with the subclass route:
```
SamplingAndMetricTrackingFnDataReceiver extends MetricTrackingFnDataReceiver {

public void accept(WindowedValue<T> value) {
  sample(value);
  super.accept(value);
}
```

Or take a look at the profiling information. If you follow the instructions here to install cloud profiler locally
https://github.com/apache/beam/blob/b64526a3a2429d30535d5cd52d99521878dd3d2c/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L1549
and then authenticate with gcloud you can run (replacing the GCP project with one that you have access to) will upload the results to Google Cloud Profiler:
./gradlew -info :sdks:java:harness:jmh:jmh -Pbenchmark=ProcessBundleBenchmark.testLargeBundle -PgcpProject=...

Reply:
I think the problem was that I was measuring the null hypothesis at upstream head, instead of origin head. I re-ran the benchmark and saw that the benchmark at origin is the same as data sampling enabled. Maybe there were some changes that resulted in some performance differences?

**yes sampling (with atomics)**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1584.137 ± 18.961  ops/s
:sdks:java:harness:jmh:jmh (Thread[included builds,5,main]) completed. Took 5 mins 11.907 secs.

**no data sampling**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1699.238 ± 16.382  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 4,5,main]) completed. Took 5 mins 11.424 secs.

**at upstream head**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1737.774 ± 34.448  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 4,5,main]) completed. Took 5 mins 8.815 secs.

**at origin head**
Benchmark                                Mode  Cnt     Score    Error  Units
ProcessBundleBenchmark.testLargeBundle  thrpt   15  1698.047 ± 17.290  ops/s
:sdks:java:harness:jmh:jmh (Thread[Execution worker Thread 4,5,main]) completed. Took 5 mins 10.494 secs.

Reply:
good to know and yes there were some performance optimizations that I added on these code paths recently

Reply:
I also ran with the profiling enabled but wasn't able to identify any more points for improvement.